File: constructTreeFromInorderPreorder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class constructTreeFromInorderPreorder {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length == 0 || inorder.length == 0) return null;
        int idx = -1;
        TreeNode root = new TreeNode(preorder[0]);
        for (int i = 0; i < inorder.length; i++) {
            if (preorder[0] == inorder[i]) {
                idx = i;
                break;
            }
        }
        int[] inLeft = Arrays.copyOfRange(inorder, 0, idx);
        int[] inRight = Arrays.copyOfRange(inorder, idx + 1, inorder.length);
        int[] preLeft = Arrays.copyOfRange(preorder, 1, inLeft.length + 1);
        int[] preRight = Arrays.copyOfRange(preorder, inLeft.length + 1, preorder.length);
        root.right = buildTree(preRight, inRight);
        root.left = buildTree(preLeft, inLeft);
        return root;
    }
}

File: ConstructTreefromInorderPreorder.py
#https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/description/
#time :O(n)
#space: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder or not inorder:
            return 

        root = TreeNode(preorder[0])
        #we know that root will be at middle in inorder 
        mid = inorder.index(preorder[0])

        #constructing left tree 
        root.left = self.buildTree(preorder[1:mid+1], inorder[:mid])

        #constructing right sub tree
        root.right = self.buildTree(preorder[mid+1:], inorder[mid+1:])

        return root

